util/sha256x/sha256.go

```diff
@@ -2,79 +2,92 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// Package sha256x is like crypto/sha256 with extra methods.
-// It exports a concrete Hash type
-// rather than only returning an interface implementation.
-package sha256x
+// Package hashx provides a concrete implementation of hash.Hash
+// that operates on a particular block size.
+package hashx
 
 import (
-	"crypto/sha256"
 	"encoding/binary"
+	"fmt"
 	"hash"
 	"unsafe"
 )
 
-var _ hash.Hash = (*Hash)(nil)
-
-// Hash is a hash.Hash for SHA-256,
-// but has efficient methods for hashing fixed-width integers.
-type Hash struct {
-	// The optimization is to maintain our own block and
-	// only call h.Write with entire blocks.
-	// This avoids double-copying of buffers within sha256.digest itself.
-	// However, it does mean that sha256.digest.x goes unused,
-	// which is a waste of 64B.
-
-	// H is the underlying hash.Hash.
-	// The hash.Hash.BlockSize must be equal to sha256.BlockSize.
-	// It is exported only for testing purposes.
-	H  hash.Hash              // usually a *sha256.digest
-	x  [sha256.BlockSize]byte // equivalent to sha256.digest.x
-	nx int                    // equivalent to sha256.digest.nx
+var _ hash.Hash = (*Block512)(nil)
+
+// Block512 wraps a hash.Hash for functions that operate on 512-bit block sizes.
+// It has efficient methods for hashing fixed-width integers.
+type Block512 struct {
+	// Hash should use a hashing algorithm that operates on 512-bit block sizes
+	// or a block size that evenly multiplies up to 512-bits.
+	// The hash still operates correctly even with misaligned block sizes,
+	// but operates less efficiently.
+	//
+	// Example algorithms with 512-bit block sizes include:
+	//   - MD4 (golang.org/x/crypto/md4)
+	//   - MD5 (crypto/md5)
+	//   - BLAKE2s (golang.org/x/crypto/blake2s)
+	//   - BLAKE3
+	//   - RIPEMD (golang.org/x/crypto/ripemd160)
+	//   - SHA-0
+	//   - SHA-1 (crypto/sha1)
+	//   - SHA-2 (crypto/sha256)
+	//   - Whirlpool
+	//
+	// See https://en.wikipedia.org/wiki/Comparison_of_cryptographic_hash_functions#Parameters
+	// for a list of hash functions and their block sizes.
+	//
+	// Block512 assumes that hash.Hash.Write never fails and
+	// never allows the provided buffer to escape.
+	hash.Hash
+
+	x  [512 / 8]byte
+	nx int
 }
 
-func New() *Hash {
-	return &Hash{H: sha256.New()}
+func New(h hash.Hash) (*Block512, error) {
+	b := &Block512{Hash: h}
+	if len(b.x)%h.BlockSize() != 0 {
+		return b, fmt.Errorf("hashx.Block512: inefficient use of hash.Hash with %d-bit block size", 8*h.BlockSize())
```

Comment:
So that's an error? But not an error?

Docs on this function?

Above it says:

```
// The hash still operates correctly even with misaligned block sizes,
// but operates less efficiently.
```

So why does this return an error?

Reply:
I don't feel strongly whether this should return an error. The only reason to use `hashx` is for performance, so I figured it would be a good sanity check that someone was not passing misaligned algorithms.

Reply:
Do the latest set of changes look alright? Thanks!